[fuse] 뉴스스탠드 #2 PR

silvertae opened the pull request:
## 미리보기
![뉴스스탠드 #2 그리드](https://github.com/codesquad-members-2023/fe-max--newsstand/assets/76121068/9d3c7a91-8e69-42d2-8610-41226ea76469)

## 이번주 주요 개발

- [x] 헤더
  - [x] 로고 클릭 시 페이지 새로고침
  - [x] 시스템 날짜 표시
- [x] 자동롤링영역
  - [x] 롤링 애니메이션
  - [x] 좌우 롤링 영역이 1초의 시간차를 두고 독립적으로 돌아가게 하기
  - [x] 좌우 롤링 영역이 5초마다 변경되는 기능
  - [x] 롤링 영역 마우스 오버시 자동롤링 멈추기 기능
- [ ] 그리드 UI
  - [x] 페이지 새로고침 시 랜덤한 순서로 이미지 표시
  - [x] 페이지 이동 기능
  - [x] 각 영역에 마우스 오버시 구독하기 버튼 나타나는 기능
  - [ ] 구독하기 버튼 클릭 시 구독된 언론사 항목 저장
  - [ ] 내가 구독한 언론사 보기 탭 구현
- [x] express 서버 기반 비동기 데이터 통신

## 고민한 부분
- Store와 View를 어떻게 나누어야 할까
- 각 컴포넌트에서 본인 상태를 가지고 관리하게 하려면 어떻게 만들어야 할까
- 부모 컴포넌트의 상태가 변경된 것을 반영해서 자식 컴포넌트를 업데이트 하려면 어떻게 해야할까
- `Object is possibly 'undefined'.ts(2532)` 오류를 없애려면 어떻게 할까

노션 링크
https://www.notion.so/2a20f362fac04192ba0353d9846bc40a?pvs=4

silvertae committed:
chore: express 패키지 Dependencies에 설치

silvertae committed:
feat: 그리드 보기에 필요한 데이터를 크롤링하는 코드 추가

- 그리드 뷰에 사용할 로고들을 필요한 개수만 가져오기 위해서 네이버에서
puppeteer로 데이터를 크롤링하는 코드 추가
- gridData.json 이라는 이름으로 새로운 데이터 추가

silvertae committed:
chore: gridData src폴더에 추가 및 기존 포맷 수정

silvertae committed:
feat: State 타입 선언 및 헤더 시스템 날짜 기능 추가

- state를 main.ts에서 json 데이터를 이용해 만들어두고 최상위 컴포넌트인
NewsStand에 props로 넘겨주도록 만들었다.
- 필요한 타입들을 만들고 types.ts 파일로 분리하였다.
- 헤더에 state에서 받아온 시스템 날짜를 이용해 화면에 렌더링 하는 기능을
추가하였다.

silvertae committed:
feat: 자동롤링 영역에 왼쪽, 오른쪽 다른 데이터를 전달하는 기능 추가

- 총 10개의 뉴스데이터에서 왼쪽에는 홀수 번째 헤드라인들을, 오른쪽에는
짝수 번째 헤드라인들을 props로 전달해서 NewsRoller 컴포넌트를 만들도록
구상했다.

silvertae committed:
feat: 3초마다 넘어가는 뉴스롤러 기능 구현

- 우선 왼쪽, 오른쪽 시간차 없이 3초마다 넘어가도록 구현
- NewsRoller 컴포넌트 state에 다음에 보여질 뉴스기사 index 추가
- @keyframes로 css 애니메이션 적용
- 타입 에러 해결 필요 (ts2532)

bakhacode commented on src/components/NewsStand.ts:
```diff
@@ -1,20 +1,31 @@
   main: Main;
 
-  constructor(props) {
+  constructor({
+    systemDate,
+    oneLineNews,
+    gridData,
+    listData,
+  }: {
+    systemDate: Date;
+    oneLineNews: OneLineNews[];
+    gridData: GridNewsData[];
+    listData: ListNewsData[];
+  }) {
```
this.systemDate 이렇게 안쓸 수 있어서 좋네용 👍

crongro commented:
생성자는 작게 유지~ 메서드로 분리하세요~ 2-3개 메서드를 더 만들수 있어보여요. 
그래야 메서드를 별도로 테스트하기도 좋고, 코드 읽기도 좋아지고요.

매직넘버 찾아서 제거해보세요.

DOM조작은 template literal 도 활용하면 더 보기 좋고 편해요.

bakhacode commented on src/components/NewsRoller.ts:
```diff
@@ -1,22 +1,65 @@
+      nextTitle.setAttribute('style', 'top: 1rem');
+      nextTitle.textContent = this.props[this.state.titleIdx].title;
+
+      this.state.titleIdx += 1;
+      this.state.titleIdx %= 5;
```
함수로 어떤 동작인지 한 눈에 읽히면 좋을 것 같아요

silvertae replied:
아직 뉴스 롤러 쪽을 구현중이다 보니 어떤 역할을 하는 함수인지 분리도 안 되어있고 매직넘버들이 있어서 더욱 그런 것 같네요. 말씀해주신 부분을 신경써서 고쳐보겠습니다.

bakhacode commented on src/components/NewsRoller.ts:
```diff
@@ -1,22 +1,65 @@
+
+      this.titleBox.append(nextTitle);
 
-    const headLine = document.createElement('div');
-    this.title = headLine;
-    headLine.classList.add('news-oneline');
-    headLine.textContent = props.oneLineNews[0].title;
+      this.currentTitle.classList.add('roll');
+      nextTitle.classList.add('roll');
 
-    this.element.append(name, headLine);
+      setTimeout(() => {
+        this.titleBox.removeChild(this.currentTitle);
+        this.currentTitle = nextTitle;
+        console.log(this.currentTitle);
+        nextTitle.removeAttribute('style');
+        nextTitle.classList.remove('roll');
+      }, 600);
+    }, 3000);
```
store랑 view 분리해서 view에서 셋인터벌의 내부 동작이 끝나고 다음 셋인터벌의 동작이 시작되기 전에 마지막 변경은 간접적으로 거쳐서 하면 좋을 것 같습니다

silvertae replied:
지금은 컴포넌트 내부 메서드 직접적으로 롤러영역을 동작시키고 있는데, 추후 flux 패턴등을 적용해서 리팩터링 해보려고 합니다. 그 때 store와 view 역할을 잘 나누어봐야겠네요.

silvertae committed:
feat: 좌우 롤링 영역 시간차와 마우스오버 시 멈추는 기능 추가

- constructor에서 받은 string으로 좌우 영역을 구분해서,
오른쪽은 처음에 1초 늦게 시작하도록 설정했다.
- setEvent()로 마우스 관련 이벤트를 설정했다. 좌우 뉴스롤러 컴포넌트는
독립적으로 동작한다.

silvertae committed:
feat: @type/express, ts-node 설치 및 서버 연결 세팅

- typescript 환경에서 express 설치하고 구동하기 위해서 필요한 패키지들을
설치하였다.
- tsconfig 파일에 필요한 설정을 추가하였다.

silvertae committed:
feat: express 서버에서 필요한 뉴스데이터를 받아오는 기능 추가

- cors 오류 해결을 위해 cors, @types/cors 패키지 설치
- 서버와 클라이언트를 동시에 동작시키기 위해서 concurrently 패키지 설치
- npm start로 한번에 실행할 수 있는 스크립트 추가
- 비동기로 데이터를 fetch 해오도록 만들려고 async/await를 활용한 함수를
만들었다.

silvertae committed:
chore: Vite 환경 변수 설정하기

- Vite 환경에서 환경변수로 서버 주소를 변수로 만들어 사용하기

silvertae committed:
feat: 그리드뷰 언론사 이미지 랜덤으로 불러오는 기능 추가

- 피셔에이츠 알고리즘을 사용해서 처음 fetch 해온 데이터로 초기 state를
설정할 때 이미지 순서를 무작위로 섞어서 불러온 뒤 하위 컴포넌트에 전달
하도록 만들었다.

silvertae committed:
feat: 그리드 뷰 페이지 이동 기능 추가

- GridView 컴포넌트에서 현재 페이지 번호를 state로 가지고 있도록 하고,
이를 변경시키면서 보여주는 언론사 이미지를 바꾸어 렌더링 하도록 만들었다.
- 좌우 화살표 버튼에 이벤트를 등록해서 페이지를 이동할 수 있게 했다.
- 첫 페이지와 마지막 페이지에서 disabled 되는 화살표는 'hide' 클래스를
주는 방식으로 display: none으로 숨김 처리했다.

silvertae committed:
feat: 각 그리드 셀 마우스오버 구독하기 버튼 나타나는 기능 추가

- 각 그리드 셀들에 event를 등록해서 마우스 오버시 새로운 구독 버튼 요소가
append 되도록 구현했다.
- 기존 언론사 이미지는 display: none으로 숨김 처리한다.
- 마우스가 영역을 떠날 시 반대로 버튼은 사라지고 언론사 이미지가 나타난다.
- 타입 오류 해결 필요

silvertae committed:
feat: 헤더 로고 클릭시 새로고침 기능 추가
